### PR TITLE
[Feature] Add DUTS dataset

### PR DESCRIPTION
**Pull request by acdart**
Support [DUTS](http://saliencydetection.net/duts/) dataset, which is the most popular dataset in salient object detection.

DUTS contains 10,553 training images and 5,019 test images. All training images are collected from the ImageNet DET training/val sets, while test images are collected from the ImageNet DET test set and the SUN data set. Both the training and test set contain very challenging scenarios for saliency detection. Accurate pixel-level ground truths are manually annotated by 50 subjects.

DUTS is currently the largest saliency detection benchmark with the explicit training/test evaluation protocol. For fair comparison in the future research, the training set of DUTS serves as a good candidate for learning DNNs, while the test set and other public data sets can be used for evaluation.

### CODE REVIEW

**Comment by MengzhangLI** [comment]
Hi, thanks for your nice pr, we would review it as soon as possible.

Salient object detection is an important branch in semantic segmentation, could you also join us in supporting other representative datasets, such as MSRA, THUR15K, ECSSD and so on? 

Best,

**Comment by Junjun2016** [comment]
[PiCANet](https://openaccess.thecvf.com/content_cvpr_2018/CameraReady/1251.pdf)
image
image

**Comment by Junjun2016** [comment]
Paper: [U^2Net](https://arxiv.org/pdf/2005.09007.pdf).
Code: [link1](https://github.com/xuebinqin/U-2-Net).
image
image
image
image
image

**Comment by Junjun2016** [comment]
U^2Net results:
Screenshot from 2021-09-28 17-40-59

**Comment by Junjun2016** [comment]
Paper: [BASNet](https://openaccess.thecvf.com/content_CVPR_2019/papers/Qin_BASNet_Boundary-Aware_Salient_Object_Detection_CVPR_2019_paper.pdf).
Code: [link1](https://github.com/xuebinqin/BASNet).

**Comment by Junjun2016** [comment]
Paper: [Cascaded Partial Decoder for Fast and Accurate Salient Object Detection (CPD CVPR'19)](https://openaccess.thecvf.com/content_CVPR_2019/papers/Wu_Cascaded_Partial_Decoder_for_Fast_and_Accurate_Salient_Object_Detection_CVPR_2019_paper.pdf).
Code: [link1](https://github.com/wuzhe71/CPD).
image
image

**Comment by Junjun2016** [comment]
Paper: [Stacked Cross Refinement Network for Edge-Aware Salient Object Detection (SCRN ICCV'19)](https://openaccess.thecvf.com/content_ICCV_2019/papers/Wu_Stacked_Cross_Refinement_Network_for_Edge-Aware_Salient_Object_Detection_ICCV_2019_paper.pdf).
Code: [link1](https://github.com/wuzhe71/SCRN).
image
image
image

**Comment by acdart** [comment]
abot metrics in SOD
1. in SOD, metrics are always computed like 'mean(calc_metric(pred[i], gt[i]))', it's different with MMSegmentation
2. need to add support for some popular metrics: MAE, max-Fmeasure, mean-Fmeasure, Smeasure, Emeasure
3. score-map output is needed by some metrics like max-Fmeasure

**Comment by Junjun2016** [comment]
> abot metrics in SOD
> 
> 1. in SOD, metrics are always computed like 'mean(calc_metric(pred[i], gt[i]))', it's different with MMSegmentation
> 2. need to add support for some popular metrics: MAE, max-Fmeasure, mean-Fmeasure, Smeasure, Emeasure
> 3. score-map output is needed by some metrics like max-Fmeasure

- add a `return_logit `argument to the [forward of segmentor](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/base.py#L97) the same as `return_loss` , pass the argument by kwargs, so we only should modify [simple_test](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py#L254), [aug_test](https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/encoder_decoder.py#L267), and eval_hooks.
- modify the calculation logic of `pre_eval_to_metrics` to support sample average metrics.
image

- integrate this [PR](https://github.com/open-mmlab/mmsegmentation/pull/295) to support reducing background class or not for BCE, especially for one channel case.

**Comment by Junjun2016** [comment]
Hi @shinya7y 
Could you please help us review this PR together?

**Comment by shinya7y** [comment]
Sure. I will only check superficial issues because I'm not familiar with SOD and mmseg.

File: docs/dataset_prepare.md [hunk 1]
```diff
@@ -253,3 +275,49 @@ Since we only support test models on this dataset, you may only download [the va
...
+
+In salient object detection (SOD), HKU-IS is used for evaluation.
+
+First，download [HKU-IS.rar](https://sites.google.com/site/ligb86/mdfsaliency/).
+
+To convert HKU-IS dataset to MMSegmentation format, you should run the following command:
+
+```shell
+python tools/convert_datasets/hku_is.py /path/to/HKU-IS.rar
+```
```
Comment (Junjun2016) [inline docs/dataset_prepare.md:323]: Could modify the Chinese document accordingly.

File: configs/_base_/datasets/duts.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+data = dict(
+    samples_per_gpu=4,
+    workers_per_gpu=4,
+    train=dict(
+        type=dataset_type,
+        data_root=data_root,
+        img_dir='images/training',
+        ann_dir='annotations/training',
+        pipeline=train_pipeline),
+    val=dict(
```
Comment (Junjun2016) [inline configs/_base_/datasets/duts.py:45]: Can we use `concat dataset` (https://github.com/open-mmlab/mmsegmentation/pull/833) and do evaluation seperately.

File: mmseg/apis/test.py [hunk 3]
```diff
@@ -38,6 +38,7 @@ def single_gpu_test(model,
                     efficient_test=False,
                     opacity=0.5,
                     pre_eval=False,
+                    return_logit=False,
```
Comment (Junjun2016) [inline mmseg/apis/test.py:41]: docstring for this new argument.

File: mmseg/models/segmentors/encoder_decoder.py [hunk 4]
```diff
@@ -216,7 +216,7 @@ def whole_inference(self, img, img_meta, rescale):
 
         return seg_logit
 
-    def inference(self, img, img_meta, rescale):
+    def inference(self, img, img_meta, rescale, return_logit):
```
[the remark this answers is not in the document]
Comment (Junjun2016) [reply mmseg/models/segmentors/encoder_decoder.py:219]: Docstring for new argument.

File: mmseg/models/segmentors/encoder_decoder.py [hunk 5]
```diff
@@ -239,7 +239,13 @@ def inference(self, img, img_meta, rescale):
             seg_logit = self.slide_inference(img, img_meta, rescale)
         else:
             seg_logit = self.whole_inference(img, img_meta, rescale)
-        output = F.softmax(seg_logit, dim=1)
+        if return_logit:
+            output = seg_logit
+        else:
+            if seg_logit.shape[1] >= 2:
```
Comment (Junjun2016) [inline mmseg/models/segmentors/encoder_decoder.py:245]: It would be better to add some comments for different cases.

**Comment by RockeyCoss** [comment]
Please merge the master branch into your branch, thank you.

**Comment by MengzhangLI** [comment]
Hi, @acdart . Sorry to bother you.

Do you plan to continue finishing this pr? We can spare more time or human resources in the next month if you need it.

Looking forward to your reply. 

Best,

**Comment by Junjun2016** [comment]
Add class name and palette to `mmseg/core/evaluation/class_names.py`.
